### PR TITLE
(0.90.2) Refactor `Fields.interpolate` to make sense for `Nothing` locations and `Flat` directions

### PR DESCRIPTION
**Pull request by glwagner**
This PR changes the syntax for interpolate to

```julia
interpolate(to_node, from_field, from_location, from_grid)
```

to interpolate `from_field` `to_node`.

Then to_node may be a 3-tuple for 3D grid/locations, 2-tuple for 2D grids/locations, or 1-tuple for 1D grids/locations.

It also makes `LatitudeLongitudeGrid` support `Flat` vertical directions.

Closes https://github.com/CliMA/Oceananigans.jl/issues/3356

### CODE REVIEW

File: src/Fields/interpolate.jl [hunk 1]
```diff
@@ -127,71 +148,150 @@ Convert the coordinates `(x, y, z)` to _fractional_ indices on a regular rectili
...
+end
 
-"""
-    interpolate(field, x, y, z)
+@inline function _fractional_indices((x, y), grid, ℓx, ℓy, ::Nothing)
+    ii = fractional_x_index(x, (ℓx, ℓy, nothing), grid)
+    jj = fractional_y_index(y, (ℓx, ℓy, nothing), grid)
+    if y > 0
+        @show (x, y)
+    end
```
Comment (simone-silvestri) [inline src/Fields/interpolate.jl:185]:
```suggestion
```

File: src/Fields/interpolate.jl [hunk 2]
```diff
@@ -127,71 +148,150 @@ Convert the coordinates `(x, y, z)` to _fractional_ indices on a regular rectili
...
+    iz = interpolator(kk)
+
+    return _interpolate(from_field, ix, iy, iz)
+end
+
+"""
+    interpolator(ii)
+
+Given the ``interpolator tuple'' for the ``fractional'' index `ii`
+defined as the 3-tuple
```
Comment (simone-silvestri) [inline src/Fields/interpolate.jl:237]:
```suggestion
Returns the ``interpolator tuple'' from the ``fractional'' index `ii`
defined as the 3-tuple
```

File: src/Grids/latitude_longitude_grid.jl [hunk 3]
```diff
@@ -237,41 +240,46 @@ function with_precomputed_metrics(grid)
...
         TZ = Bounded
     end
 
+    # Validate longitude and latitude
+    λ₁, λ₂ = get_domain_extent(longitude, Nλ)
+    λ₂ - λ₁ ≤ 360 || throw(ArgumentError("Longitudinal extent cannot be greater than 360 degrees."))
+    λ₁ <= λ₂      || throw(ArgumentError("Longitudes must increase west to east."))
+
+    φ₁, φ₂ = get_domain_extent(latitude, Nφ)
+    -90 <= φ₁ || throw(ArgumentError("The southernmost latitude cannot be less tha -90 degrees."))
```
Comment (simone-silvestri) [inline src/Grids/latitude_longitude_grid.jl:264]:
```suggestion
    -90 <= φ₁ || throw(ArgumentError("The southernmost latitude cannot be less than -90 degrees."))
```

File: src/Grids/latitude_longitude_grid.jl [hunk 4]
```diff
@@ -237,41 +240,46 @@ function with_precomputed_metrics(grid)
...
+    φ₂ <= 90  || throw(ArgumentError("The northern latitude cannot be less tha -90 degrees."))
```
Comment (simone-silvestri) [inline src/Grids/latitude_longitude_grid.jl:265]:
```suggestion
    φ₂ <= 90  || throw(ArgumentError("The northern latitude cannot be less than -90 degrees."))
```

File: src/Grids/latitude_longitude_grid.jl [hunk 5]
```diff
@@ -237,41 +240,46 @@ function with_precomputed_metrics(grid)
...
+    z         = validate_dimension_specification(TZ, z,         :z,         Nz, FT)
 
-    longitude = validate_dimension_specification(TX, longitude, :x, Nλ, FT)
-    latitude  = validate_dimension_specification(TY, latitude,  :y, Nφ, FT)
-    z         = validate_dimension_specification(TZ, z,         :z, Nz, FT)
+    halo = validate_halo(TX, TY, TZ, halo)
+    topology = (TX, TY, TZ)
 
-    return Nλ, Nφ, Nz, Hλ, Hφ, Hz, latitude, longitude, z, (TX, TY, TZ), precompute_metrics
+    return topology, size, halo, latitude, longitude, z, precompute_metrics
```
Comment (simone-silvestri) [inline src/Grids/latitude_longitude_grid.jl:282]:
```suggestion
    return TX, TY, TZ, size, halo, longitude, latitude, z, precompute_metrics
```
to follow
https://github.com/CliMA/Oceananigans.jl/blob/8c42e9b127e7d07220a65051e0909122c8175086/src/Grids/rectilinear_grid.jl#L297?

**Review by simone-silvestri — APPROVED** [review]

File: src/Grids/latitude_longitude_grid.jl (hunk 5 above)
Comment (glwagner) [reply src/Grids/latitude_longitude_grid.jl:282]: I feel its easier to read with fewer return args, I will change `rectilinear_grid` accordingly

**Comment by glwagner** [comment]
@simone-silvestri this has some changes to support `Flat` directions for lat-lon so I think we want to get this over the line before #3343

File: Project.toml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 name = "Oceananigans"
 uuid = "9e8cae18-63c1-5223-a75c-80ca9d6e9a09"
 authors = ["Climate Modeling Alliance and contributors"]
-version = "0.90.0"
+version = "0.90.1"
```
Comment (navidcy) [inline Project.toml:4]:
```suggestion
version = "0.90.2"
```

File: Manifest.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # This file is machine-generated - editing it directly is not advised
 
-julia_version = "1.9.3"
+julia_version = "1.9.2"
```
Comment (navidcy) [inline Manifest.toml:3]: why are there Manifest changes? any updates?
if so, let's use v1.9.3?

**Comment by glwagner** [comment]
@simone-silvestri so close! Can you take a look at the distributed tests?

**Comment by glwagner** [comment]
Superceded by #3395

**Pull request closed without merging**